ocaml-lsp-server/src/dune.ml

```diff
@@ -233,43 +250,98 @@ end = struct
               let+ () = Progress.build_progress progress p in
               Some ()))
 
-  let diagnostic_loop client running dune_diagnostic_id diagnostics
-      ~include_promotions =
+  let registration_of_promotion (d : Drpc.Diagnostic.Promotion.t) =
+    let registrations =
+      let in_source = Drpc.Diagnostic.Promotion.in_source d in
+      let code_action =
+        let id = "ocamllsp-promote-action/" ^ in_source in
```

Comment:
1. OK.
2. What do you mean as a capability? We register each file because we're telling the editor that we can offer code actions for that file.

Reply:
I meant that why it works differently from any other code action. The code action could be sent whenever a code action in a suitable file was triggered regardless of cursor position in the code action request.

Reply:
The reason is because we're offering code actions for files the LSP server hasn't registered for listening. By default, the lsp server is only queried for code actions for files that it will get Open/Close/Edit notifications. We need to tell the server that other files (non reason, dune, menhir, etc) may also have code actions available.